keps/sig-storage/177-volume-snapshot-validation/tighten-validation-webhook-crd.md

```diff
@@ -0,0 +1,799 @@
+- Corresponding admission controllers(MutatingAdmissionWebhook, ValidatingAdmissionWebhook) is enabled. (in v1.18+, both will be enabled by default, with mutating precedes validating)
+- API admissionregistration.k8s.io/v1beta1 or admissionregistration.k8s.io/v1 is enabled.(Prefer v1 over v1beta1)
+
+Admission controllers have been in common use in kubernetes for a long time. Admission webhooks are the new, preferred way to control admission, especially for external (out-of-tree) components like the CSI external snapshotter.
+
+A webhook server receives AdmissionReview(definition) requests from API server, and responses with a response of the same type to either admit/deny the request. Following simplified diagram shows the workflow.
+(Note that the mutatting webhooks will be invoked BEFORE validating).
+
+The webhook server will expose an HTTP endpoint such that to allow the API server to send AdmissionReview requests. Webhook server providers can dynamically(details) configure what type of resources and what type of admission webhooks via creating CRs of type ValidatingWebhookConfiguration and/or MutatingWebhookConfiguration. It’s worth thinking to provide a generic webhook server in csi-repo. At this stage, the proposal is to create a repo under CSI org, however only implement the validating webhooks for VolumeSnapshot.
+
+CRD validation is preferred over webhook validation due to their lower complexity, however CRD validation schema is unable to enforce immutability or provide ratcheting validation.
+
+### Goals
+
+- Provide an updated CRD schema to validate fields
+- Prevent:
+  - Invalid VolumeSnapshot/VolumeSnapshotContent from creation and update
+  - Invalid updates on immutable fields, i.e., VolumeSnapshot.Spec.Source
+- Provide a pre-built image which can be used to deploy the server
+- Provide a way to deploy the webhook server in cluster
+- Provide a way to authenticate the webhook server to the API server via TLS
+- Provide a release process to safely tighten the validation and move towards the ideal state of using builtin CRD validation while maintaining backwards compatibility
+
+### Non-Goals
+
+- Provide a way to authenticate the API server to the webhook server
+
+## Proposal
+
+Tighten the validation on Volume Snapshot objects. The following fields will begin to enforce immutability: `VolumeSnapshot.Spec.Source` and `VolumeSnapshotContent.Spec.Source`. The following fields will begin to enforce `oneOf`: `VolumeSnapshot.Spec.Source` and `VolumeSnapshotContent.Spec.Source`). The following fields will begin to enforce non-empty strings: `spec.VolumeSnapshotClassName`. More details are in the Validating Scenarios section.
+
+Due to backwards compatibility concerns, the tightening will occur in two phases.
+
+1. The first phase is webhook-only, and will use ratcheting validation combined with a reconciliation method to delete or fix currently persisted objects which are invalid under the new (strict) validation rules.
```

Comment:
I'm concerned about automatically deleting objects (even if they are invalid), and users or controllers being surprised/confused that their object is gone.

Is there any precedent for doing this?

Reply:
What if the original controller did nothing with the invalid objects?

Reply:
There are couple of rounds of discussion around the automated deletion/fixing for existing invalid VSC/VS resources in API server. A preferred proposal ATM is:
1. ratcheting validation using webhook such that no more invalid VSC/VS will be created in the first place, AND prevent updating of a valid VS/VSC to an invalid one.
2. In the controller side, not trying to delete/fix invalid existing CRs to avoid potential data loss, however, log error messages, and update VS/VSC status with a proper error (snapshot controller is doing that as of today, might need to have a more explicit message to state that further upgrading of the controller + CRD WILL cause the invalid CRs not deletable if the user does not take action on them). Also, we should make an announcement document the changes and leave a graceful period for users to clean up. Considering the amount of usage of volume snapshot feature at this moment, this seems to be a reasonable approach.
3. with a new release phrase when strict validation will be in-place, users SHOULD be able to downgrade to a previous version to clean invalid CRs if needed. and this process should be documented and tested.